Add blog post on Service Weaver Kubernetes deployers (#675)
ServiceWeaver/weaver: website/blog/kubernetes.md

generates Kubernetes resources that enable the application to run in a Kubernetes
cluster.

![Kube Overview](deployers/kube/kube_overview.png)

>With `Kube`:
>* The Kubernetes manifest file to deploy your application is automatically
generated.
>* You control how to run your application (e.g., resource requirements, scaling
>  specifications, volumes).
>* You decide how to export telemetry (e.g., traces to [Jaeger][jaeger], metrics
  to [Prometheus][prometheus], write custom telemetry plugins to export telemetry to
  your favorite observability framework).
>* You can use existing tools to deploy your application (e.g., [kubectl][kubectl],
  CI/CD pipelines like [GitHub Actions][github_actions], [Argo CD][argocd] or
  [Jenkins][jenkins]).

<h3 class="emphasize-title">Hello World!</h3>

To deploy a "Hello, World!" Service Weaver application with the `Kube` deployer,
the user has to write an application config and a deployment config:

```toml
# app_config.toml
[serviceweaver]
binary = "./hello"

# Per component config.
...
```

```yaml
# dep_config.yaml
appConfig: app_config.toml
repo: docker.io/mydockerid

listeners:
  - name: hello
    public: true
```

> Then use the `weaver kube deploy` to generate the Kubernetes resources needed to

Reviewer: ```suggestion
Then use `weaver kube deploy` to generate the Kubernetes resources needed to
```

Author: Done

> run your application in `/tmp/kube_ffa65856.yaml`:

Reviewer: ```suggestion
run your application:
```

Author: Done

```console
$ go build .
$ weaver kube deploy dep_config.yaml
...
Building image hello:ffa65856...
...
Uploading image to docker.io/mydockerid/...
...
Generating kube deployment info ...
...
kube deployment information successfully generated
/tmp/kube_ffa65856.yaml
```

Finally, you can simply deploy your application:

```console
$ kubectl apply -f /tmp/kube_ffa65856.yaml
```

<h3 class="emphasize-title">Configurations</h3>

With `Kube` the user can control how the application is running on Kubernetes. We

> identified the top Kubernetes knobs that the users typically configure when

Reviewer: ```suggestion
identified the top Kubernetes knobs that users typically configure when
```

Author: Done

running on Kubernetes (based on a survey done by the Go team in Q2 2023), and
expose them in the config. We expose the following knobs:

* [Resources][kubernetes_res] needed to run each container; e.g., the CPU/Mem requests/limits.
* [Scaling specifications][kubernetes_scaling] to control how the pods are scaled up/down; e.g., min and max replicas, metrics to scale, utilization threshold.
* [Volumes][kubernetes_volumes] to mount storage, config maps, secrets, etc.
* [Probes][kubernetes_probes] to monitor the liveness, readiness and startup of the containers.

**Note** that these knobs can be specified for the entire application, or for
a subset of the components.

For example, if the user wants to configure the memory limits for all the running pods,

> it can configure as follows:

Reviewer: ```suggestion
it can be configured as follows:
```

Author: Done

```yaml
appConfig: weaver.toml
repo: docker.io/mydockerid

listeners:
- name: hello
  public: true

resourceSpec:
  requests:
    memory: "64Mi"
```

**Note** that if there are other knobs that you want to configure, you can manually
edit the generated Kubernetes manifest file. However, if there are knobs that are

> widely used, we can expose them along the existing knobs in the config.

Reviewer: ```suggestion
widely used, we can expose them along with the existing knobs in the config.
```

Author: Done

<h3 class="emphasize-title">Telemetry</h3>

The [`Kube` deployer documentation][kube_doc_telemetry] goes into details how to
export telemetry to your favorite observability framework. To do that, you need
to implement a wrapper deployer on top of the `Kube` deployer using the
[Kube tool][kube_telemetry_api] abstraction. We provide examples on how to export
traces to [Jaeger][jaeger], metrics to [Prometheus][prometheus] and how to visualize
the exported traces and metrics with [Grafana][grafana].

<h3 class="emphasize-title">CI/CD Pipelines</h3>

The `Kube` deployer allows you to deploy your application using your favorite CI/CD
pipeline. [Here][github_actions] is an example on how to integrate with the

> [GitHub Actions][gha]. We also tried out integrating with [ArgCD][argocd] and

Reviewer: ```suggestion
[GitHub Actions][gha]. We also integrate with [ArgCD][argocd] and
```

Author: Done
